pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,399 @@
+// have	Openflow flows installed for cross-cluster traffic.
+func (c *MCPodRouteController) validatePod(pod *corev1.Pod) bool {
+	podIP := pod.Status.PodIP
+	podNodeIP := pod.Status.HostIP
+
+	if pod.Spec.NodeName == c.nodeConfig.Name {
+		klog.V(2).InfoS("Skip Pod on the current Node")
+		return false
+	}
+
+	if podIP == "" || podNodeIP == "" || pod.Spec.HostNetwork {
+		klog.V(2).InfoS("Pod has no valid IPs or it is in HostNetwork, skip it", "name", klog.KObj(pod),
+			"podIP", podIP, "nodeIP", podNodeIP, "hostNetwork", pod.Spec.HostNetwork)
+		return false
+	}
+
+	ips, exists := c.getPodNodeIPs(types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
+	if exists && ips.podNodeIP == podNodeIP && ips.podIP == podIP {
+		klog.V(2).InfoS("No Pod change impacts installed Openflow rules", "name", klog.KObj(pod))
+		return false
+	}
+	return true
+}
+
+func (c *MCPodRouteController) enqueuePod(obj interface{}, isDelete bool) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	c.podQueue.Add(types.NamespacedName{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	})
+}
+
+func (c *MCPodRouteController) Run(stopCh <-chan struct{}) {
+	defer c.gwQueue.ShutDown()
+	defer c.podQueue.ShutDown()
+
+	klog.InfoS("Starting controller", "controller", podRouteControllerName)
+	defer klog.InfoS("Shutting down controller", "controller", podRouteControllerName)
+	if !cache.WaitForNamedCacheSync(podRouteControllerName, stopCh, c.gwInformer.HasSynced) {
+		return
+	}
+	// Run a single routine to handle Gateway events.
+	go wait.Until(c.gatewayWorker, time.Second, stopCh)
+	<-stopCh
+}
+
+func (c *MCPodRouteController) gatewayWorker() {
+	for c.processGatewayNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processGatewayNextWorkItem() bool {
+	key, quit := c.gwQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gwQueue.Done(key)
+
+	if k, ok := key.(string); !ok {
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(k); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway(key string) error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	localStopCh := make(chan struct{})
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if activeGW == nil || c.nodeConfig.Name != activeGW.Name {
+		if c.started {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(localStopCh)
+			c.started = false
+		}
+		if len(c.podInfos) > 0 {
+			return c.deletePodFlows()
```

Comment:
Here the worker goroutines may still be running and can install more flows after `deletePodFlows()` is called. I guess we need a synchronization between workers and `syncGateway()`.